### PR TITLE
[Document provider] Rename File/Folder

### PR DESCRIPTION
**Pull request by abelgardep**
Implements #2484 

If rename succeeds -> update view
If rename fails -> sync with server to check what could have failed

____

Test Plan:

https://github.com/owncloud/QA/blob/master/Mobile/Android/Release_2.11.0/2506-document_provider_rename.md

Bugs & improvements:

- [x] (1) Submit with the same name https://github.com/owncloud/android/pull/2506#issuecomment-482015425

### CODE REVIEW

**Comment by davigonz** [comment]
@abelgardep can you please solve the conflicts? I will start with the code review after that.

**Comment by abelgardep** [comment]
@davigonz Done! Ready to CR

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+        sizeHint: Point?,
+        signal: CancellationSignal?
+    ): AssetFileDescriptor {
+
+        val docId = documentId.toLong()
+        updateCurrentStorageManagerIfNeeded(docId)
+
+        val file = currentStorageManager!!.getFileById(docId)
+
+        val realFile = File(file!!.storagePath)
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: Same with file

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+import java.util.HashMap
+import java.util.Vector
+
+class DocumentsStorageProvider : DocumentsProvider() {
+    /**
+     * If a directory requires to sync, it will write the id of the directory into this variable.
+     * After the sync function gets triggered again over the same directory, it will see that a sync got already
+     * triggered, and does not need to be triggered again. This way a endless loop is prevented.
+     */
+    private var requestedFolderIdForSync: Long = -1
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: This variable is only used in queryChildDocuments so it does not need to be a global variable.

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        return result
+    }
+
+    private fun toNotifyUri(uri: Uri): Uri {
+        return DocumentsContract.buildDocumentUri(
+            documentsProviderAuthority, uri.toString()
+        )
+    }
+
+    private fun toUri(documentId: String): Uri {
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: And rewrite this function in just one line 

`private fun toUri(documentId: String): Uri =  Uri.parse(documentId)`

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+            } else {
+                if (f.fileName.contains(query)) {
+                    result.add(f)
+                }
+            }
+        }
+        return result
+    }
+
+    private fun toNotifyUri(uri: Uri): Uri {
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: You could rewrite this function like this
`private fun toNotifyUri(uri: Uri): Uri = DocumentsContract.buildDocumentUri(documentsProviderAuthority, uri.toString())`

File: owncloudApp/src/main/java/com/owncloud/android/providers/cursors/FileCursor.kt [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+import android.os.Bundle
+import android.provider.DocumentsContract
+import android.provider.DocumentsContract.Document
+import com.owncloud.android.datamodel.OCFile
+import com.owncloud.android.utils.MimetypeIconUtil
+
+class FileCursor(projection: Array<String>?) : MatrixCursor(projection ?: DEFAULT_DOCUMENT_PROJECTION) {
+    private var cursorExtras = Bundle.EMPTY
+
+    override fun getExtras(): Bundle {
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/providers/cursors/FileCursor.kt]: Function in just one line

**Comment by davigonz** [comment]
@abelgardep some changes requested

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt (hunk 2 above)
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: But we need it for next calls

**Comment by abelgardep** [comment]
@davigonz Changes done

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 6]
```diff
@@ -271,27 +264,24 @@ class DocumentsStorageProvider : DocumentsProvider() {
...
-                if (f.fileName.contains(query)) {
-                    result.add(f)
-                }
+
+        val folderContent = currentStorageManager?.getFolderContent(root, false) ?: return result
+
+        for (f in folderContent) {
+            if (f.fileName.contains(query)) {
+                result.add(f)
+                if (f.isFolder) result.addAll(findFiles(f, query))
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: Are you sure this is the same logic than before? If f is folder, we are adding it as well as add all while with the previous implementation we just added all.

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 7]
```diff
@@ -271,27 +264,24 @@ class DocumentsStorageProvider : DocumentsProvider() {
...
-            if (f.isFolder) {
-                result.addAll(findFiles(f, query))
-            } else {
-                if (f.fileName.contains(query)) {
-                    result.add(f)
-                }
+
+        val folderContent = currentStorageManager?.getFolderContent(root, false) ?: return result
+
+        for (f in folderContent) {
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: what about using forEach here?

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 8]
```diff
@@ -97,6 +97,7 @@ class DocumentsStorageProvider : DocumentsProvider() {
 
         val resultCursor = FileCursor(projection)
 
+        // Create result cursor before syncing folder again, in order to enable faster loading
```
Comment (davigonz) [inline owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: ❤️

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt (hunk 6 above)
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: It is not. This way, we can look for not only files but also folders. I think it is better, what do you think?

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt (hunk 7 above)
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: Done.

**Comment by abelgardep** [comment]
@davigonz Changes done

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt (hunk 6 above)
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: Yes, understood.

**Comment by davigonz** [comment]
Code approved, ready to test @jesmrec

**Comment by jesmrec** [comment]
@abelgardep do we have control over the rename dialog?

Check these steps:

1. In Files, select to rename an item -> Rename dialog displayed
2. Submit the dialog with no changes in the item name

Current: Error `Failed to rename document`
Expected: Different options: do nothing in case the name is the same, disable the OK option... but there is any error. 

is it a document provider issue?

**Comment by abelgardep** [comment]
> @abelgardep do we have control over the rename dialog?
> 
> Check these steps:
> 
> 1. In Files, select to rename an item -> Rename dialog displayed
> 2. Submit the dialog with no changes in the item name
> 
> Current: Error `Failed to rename document`
> Expected: Different options: do nothing in case the name is the same, disable the OK option... but there is any error.
> 
> is it a document provider issue?

No, we have not.

As far as i know, Documents Provider takes care about it.

**Comment by jesmrec** [comment]
Ok, that was the only issue detected... let's trust in the future3 document provider to improve the error handling.

Approved.

**Comment by jesmrec** [comment]
@abelgardep please update the branch, so we will be able to merge and so on

**Comment by davigonz** [comment]
Thanks @abelgardep for this new feature, great job!

**Pull request merged**